meta/src/main/scala/meta/RisePrimitiveGenerator.scala

```diff
@@ -0,0 +1,268 @@
+
+}""".toString()}
+                            |""".stripMargin
+
+              os.write.over(outputPath, code)
+            case (name, _, typeSignature) =>
+              println(s"Could not generate code for `$name' as type signature `$typeSignature' is not well kinded.")
+          }
+      }
+    })
+  }
+
+  def generateObject(name: String, typeSignature: TypeAST): scala.meta.Term.Block = {
+    import scala.meta._
+    val generated = q"""{
+      object ${Term.Name{name}} extends Builder {
+        private final case class Primitive()(override val t: Type = TypePlaceholder)
+          extends rise.core.Primitive
+        {
+          override val name: String = ${Lit.String(name)}
+          override def setType(ty: Type): Primitive = Primitive()(ty)
+          override def typeScheme: Type = ${generateTypeScheme(typeSignature, Map.empty)}
+
+          override def equals(obj: Any): Boolean = obj match {
+            case p: Primitive => p.t =~~= t
+            case _ => false
+          }
+        }
+
+        override def toString: String = ${Lit.String(name)}
+        override def primitive: rise.core.Primitive = Primitive()()
+        override def apply: ToBeTyped[rise.core.Primitive] = toBeTyped(Primitive()())
+        override def unapply(arg: Expr): Boolean = arg match {
+          case _: Primitive => true
+          case _ => false
+        }
+      }
+      }"""
+    generated
+  }
+
+  def generateCaseClass(name: String, paramsString: String, typeSignature: TypeAST): scala.meta.Term.Block = {
+    import scala.meta._
+
+    val params = paramsString.split(",").map(param => {
+      val parts = param.split(":").map(_.trim)
+      param"${Term.Name(parts(0))}: ${Type.Name(parts(1))}"
+    } ).toList
+    val args: List[Term.Name] = params.map(p => Term.Name(p.name.value))
+    val types: List[Type] = params.map(p => p.decltpe.get)
+
+    val generated =
+      q"""
+      final case class ${Type.Name(name)}(..$params) extends Builder {
+        override def toString: String = ${Lit.String(name)}
+        override def primitive: rise.core.Primitive = ${Term.Name(name)}.Primitive(..$args)()
+        override def apply: ToBeTyped[rise.core.Primitive] = toBeTyped(${Term.Name(name)}.Primitive(..$args)())
+
+        override def unapply(arg: Expr): Boolean = arg match {
+          case _: Primitive => true
+          case _ => false
+        }
+      }
+
+      object ${Term.Name(name)} {
+        private final case class Primitive(..$params)(override val t: Type = TypePlaceholder)
+          extends rise.core.Primitive
+        {
+          override val name: String = ${Lit.String(name)}
+          override def setType(ty: Type): Primitive = Primitive(..$args)(ty)
+          override def typeScheme: Type = ${generateTypeScheme(typeSignature, Map.empty)}
+
+          override def equals(obj: Any): Boolean = obj match {
+            case p: Primitive =>
+              ${generateComparisonChain(args)} && p.t =~~= t
```

Comment:
Same here, replacing `p.t =~~= t` with `p.name == name` should be ok

Reply:
As above: I am just doing what the current macro in `master is doing:
https://github.com/rise-lang/shine/blob/master/macros/src/main/scala/rise/macros/Primitive.scala#L101-#L106